### ISSUE TITLE
np.array truncates string objects to 64 characters, when accessed via __array__.

### ISSUE THREAD

**Issue by konstantint** [numpy/numpy#12256]
For some reason `np.array` (or `np.asarray`) seems to truncate string objects to 64 characters, when the input argument is providing the data via an `__array__` accessor as an array of `dtype('O')` and `np.array` is asked to convert the data to `dtype('str')`.

### Reproducing code example:

```python
import numpy as np
class MyArray:
    def __array__(self, result=None):
        return np.array(['0'*70], dtype=object)
    
np.array(MyArray(), dtype=str)
```
Result (observe the truncated length of the string):
```
array(['0000000000000000000000000000000000000000000000000000000000000000'],
      dtype='<U64')
```

The practical impact of this bug is best observed in the context of Pandas:
```python
import pandas as pd
np.array(pd.Series(['0'*70]), dtype=str)
```
Result: 
```
array(['0000000000000000000000000000000000000000000000000000000000000000'],
      dtype='<U64')
```

### Numpy/Python version information:

1.14.0 3.6.3 |Anaconda, Inc.| (default, Oct 15 2017, 03:27:45) [MSC v.1900 64 bit (AMD64)]

**Comment by eric-wieser** [comment]
What an alarming bug...

Caused by these lines:
```
https://github.com/numpy/numpy/blob/master/numpy/core/src/multiarray/convert_datatype.c#L223-L292
```

**Comment by mkielar** [comment]
I can observe similar behaviour with Pandas 0.25.1 and Numpy 1.17.2, when trying to convert Shapely Linestring objects to their string representation with `astype(str)`.

In this case, this code:
```
print(edges['geometry'].dtypes)
edges['geometry'] = edges['geometry'].astype(str)
print(edges['geometry'].dtypes)
```
produces this output:
```
geometry
object
```
Also, after the conversion all strings are 64character long.
So it seems it's converting what it thinks is object type to string, but then truncates the representation to 64 chars.

We worked this around by using:
```
edges['geometry'] = edges['geometry'].apply(str)
```
which seems to work fine (however also producing `object` dtype, not S, but we can live with that).

Using:
```
edges['geometry'] = edges['geometry'].astype(np.dtype('|S4096'))
```
also worked on a test case, but this is not usable in our case, as we cannot really tell what the length of the Linestring will be after its converted to string.

**Comment by mattip** [comment]
The pandas part of the original report does not fail (pandas 0.25.3) but using a class with a custom `__array__` still does. The problem, as @eric-wieser points out, is in `PyArray_AdaptFlexibleDtype` which does not check for `__array__` on the `data_obj`

**Comment by seberg** [comment]
This was already fixed in 1.19 (or maybe even 1.18).
